https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>

a71f25789b531f8c3f3979a41d56c0b021bbe533

In [2]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score


## Part 1. Loading the dataset

In [3]:
# Using pandas load the dataset (load remotely, not locally)
path = '/workspaces/assignment-4-equation-of-a-slime-<redacted>/science_data_large.csv'
data = pd.read_csv(path)

# Output the first 15 rows of the data
print(data.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print(data.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [4]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data[['Temperature °C', 'Mols KCL']]
y = data['Size nm^3'] 

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

#confirming splits
print("Training set size (X_train):", X_train.shape)
print("Test set size (X_test):", X_test.shape)
print("Training labels size (y_train):", y_train.shape)
print("Test labels size (y_test):", y_test.shape)

Training set size (X_train): (900, 2)
Test set size (X_test): (100, 2)
Training labels size (y_train): (900,)
Test labels size (y_test): (100,)


## Part 3. Perform a Linear Regression

In [5]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report on the score for that model, in your own words (markdown, not code) explain what the score means

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)


# Create a sample datapoint and predict the output of that sample with the trained model
sample = pd.DataFrame([[500, 600]], columns=['Temperature °C', 'Mols KCL'])
predicted_size = model.predict(sample)
print(predicted_size)

#sample was 500 temp, 600 mol, output ranging between 635000 to 645000

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print(f"model R-squared score: {score}")

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")


[642323.06317814]
model R-squared score: 0.863039426921578
Coefficients: [ 870.67624734 1040.98508865]
Intercept: -417606.1136781578


The score ranges between 0.85 to 0.90 when run again. The function .score outputs the R^2 score, which basically represents "The proportion of the variance in the target variable that is predictable from the feature variables" (temp and mol). Additionally, the score ranges from 0 - 1. When the score is 1 it means that the model perfectly predicts the outcomes, when the score is 0 it means that the model predicts the outcomes the same as the mean of the data, and when the score is negative  it means that the model is worse than predicting the mean (horizontal line). In my case, my R^2 value is pretty good, it is at 0.85-0.9 which shows that our linear regression model is a fairly good fit for our data.


Sample equation: $E = mc^2$

My H(x) function: $h(x) = \theta_0 + \theta_1 x_1 + \theta_2 x_2$



$h(x) = -415385.43 + 876.40 \cdot \text{Temperature} + 1027.13 \cdot \text{Mols KCL}$



please keep in mind that these are the current values of the intercepts, when I rerun, i'll get a different intercept and slope value, but it'll be within a similar range. (I know I can use the parameter random_state=42 in my train-test split function which will ensure that I get the same training and testing data no matter how many times I run your function, But I didn't want it to be set)

Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [6]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
cross_val_scores = cross_val_score(model, X, y, cv=10)
print(cross_val_scores)

# Report on their finding and their significance
mean_score = cross_val_scores.mean()
std_score = cross_val_scores.std()

print(mean_score)
print(std_score)
# Report on their finding and their significance



[0.81123596 0.86440978 0.87808742 0.86561069 0.87495621 0.84484397
 0.87941022 0.86349411 0.78353682 0.88686516]
0.8552450341984701
0.03152876296534241


# Report on their finding and their significance
The cross-valudation score shows how well the lin regression works on different splits on the data, the scores show the proportion of the variance in the target variable. Values with 1 indicate a better fit. The mean cross-validation score is 85.5% which is pretty good. The significance of this is that on average, the model can explain 85.5% of the variance in the target variable (size) based on the features. Hence, i'd say it's a good fit. The standard deviation was approximately 0.0315, this value is pretty low. It shows that the model is quite consistent across different splits of data and hence the model can generalize pretty well and is not overly sensitive to specific parts of the dataset. The low st. dev shows that the model is not heavily overfitting or underfitting the data. Also, the cv parameter in teh cross_val_score function stands for eg. cv = 10, 10-fold cross validation. This basically means that the dataset is split into 10 equal sized subsets. Hence, the model is trained and tested 10 times, each time using a different fold as the test set. 

## Part 5. Using Polynomial Regression

In [9]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_poly_train = poly.fit_transform(X_train)
X_poly_test = poly.transform(X_test)

poly_model = LinearRegression()
poly_model.fit(X_poly_train, y_train)

y_poly_pred = poly_model.predict(X_poly_test)
poly_r2_score = r2_score(y_test, y_poly_pred)

poly_coefficients = poly_model.coef_
poly_intercept = poly_model.intercept_

print(poly_r2_score, poly_coefficients, poly_intercept)

cross_val_scores_poly = cross_val_score(poly_model, poly.fit_transform(X), y, cv=10)

mean_poly_score = cross_val_scores_poly.mean()
std_poly_score = cross_val_scores_poly.std()

(mean_poly_score, std_poly_score)

1.0 [ 0.00000000e+00  1.20000000e+01 -1.18021764e-07 -8.36308800e-12
  2.00000000e+00  2.85714287e-02] 1.617753878235817e-05


(np.float64(1.0), np.float64(0.0))

# Report on the metrics and output the resultant equation as you did in Part 3.

$h(x) = 1.617753878235817 \times 10^{-5} + 12 \cdot \text{Temperature} + (-1.18021764 \times 10^{-7}) \cdot \text{Mols KCL} + 2 \cdot (\text{Temperature} \cdot \text{Mols KCL}) + (-8.36308800 \times 10^{-12}) \cdot \text{Temperature}^2 + 0.02857 \cdot \text{Mols KCL}^2$

the polynomial regression got score of 1.0 across different splits. As mentioned earlier, the score represents the proportion of variance in the target variable which is explained by the features. A score of 1 shows a perfect fit. 

the CV resuts show the mean score as 1.0 which shows that the model fits the data across all splits perfectly. The standard deviation is 0.0, which suggests that there is no variability in the model's performance across the different splits

as with linear regression, the dataset is divided into 10 equal subsets, with the model trained and tested 10 times. the model can explain 100% of the variance in the target variable (size) based on the input (temp and mols). A perfect CV score shows that the polynomial regression is a perfect match for the data. However this might be a rpoblem of overfitting. But since we dod the crossvalidation check and it's still consistent with zero st. dev. just suggests that the model is perfect and generalizes extremely well. 